test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java

```diff
@@ -176,83 +179,177 @@ public TestCase nodes(int nodeCount)
+         * with CURRENT = v41, minimalSupportedVersion = v30, and maximalSupportedVersion = v41
+         * the test case will run:
+         * - v30 -> v41
+         * - v3x -> v41
+         * - v40 -> v41
          */
-        public TestCase upgradesTo(Semver from, Semver to)
+        public TestCase applicableVersionsRange(Semver minimalSupportedVersion, Semver maximalSupportedVersion)
         {
-            List<TestVersions> upgrade = new ArrayList<>();
-            NavigableSet<Semver> vertices = sortedVertices(SUPPORTED_UPGRADE_PATHS);
-            for (Semver start : vertices.subSet(from, true, to, false))
-            {
-                // only include pairs that are allowed, and start or end on CURRENT
-                if (SUPPORTED_UPGRADE_PATHS.hasEdge(start, to) && contains(start, to, CURRENT))
-                    upgrade.add(new TestVersions(versions.getLatest(start), Collections.singletonList(versions.getLatest(to))));
-            }
-            logger.info("Adding upgrades of\n{}", upgrade.stream().map(TestVersions::toString).collect(Collectors.joining("\n")));
-            this.upgrade.addAll(upgrade);
+            this.minimalApplicableVersion = minimalSupportedVersion;
+            this.maximalApplicableVersion = maximalSupportedVersion;
+            addUpgrades();
             return this;
         }
 
         /**
-         * performs all supported upgrade paths from the "from" target; example
-         * {@code upgradesFrom(4.0, 4.2); // produces: 4.0 -> 4.1, 4.0 -> 4.2}
-         */
-        public TestCase upgradesFrom(Semver from, Semver to)
+         * this is a deprecated, backwards-compatible API for upgrade tests, that allows to specify which upgrades to perform
+         *
+         * upgradesToCurrentFrom(Semver from) sets the upgrades to perform to be all supported direct upgrades (upgrade
+         * paths with length 1) to the CURRENT version, but omitting versions older than the given "from" version
+         *
+         * This intent is best expressed by the newer alternatives:
+         * - appplicableVersionsRange(from, CURRENT)
+         * or, perhaps, depending on the actual case:
+         * - minimalApplicableVersion(from)
+         *
+         * please note, that upgrades that do not involve the CURRENT version are not meant to be run in upgrade tests
+         *
+         * Beware, there may be no supported direct upgrade path from -> CURRENT, and this is OK.
+         * e.g. when CURRENT == v4
+         * {@code upgradesToCurrentStartingNoEarlierThan(3.0); // produces: 3.0 -> CURRENT, 3.11 -> CURRENT, …}
+         * and when CURRENT == v5
+         * {@code upgradesToCurrentStartingNoEarlierThan(3.0); // produces: 4.0 -> CURRENT, 4.1 -> CURRENT, …}
+         * and when CURRENT == v3.11
+         * {@code upgradesToCurrentStartingNoEarlierThan(3.0); // produces: 3.0 -> CURRENT, …}
+         **/
+        @Deprecated
+        public TestCase upgradesToCurrentFrom(Semver from)
```

Comment:
I retained the old API in case some test is written in an older version and then propagated to `trunk`.
I wanted such test to work out-of-the-box (with no compilation errors etc)
I don't know if this is C* way of doing things.

Reply:
in this class we don't need to go through the deprecation cycle.

it would be better to apply the patch to all branches to avoid this problem.